Delay searchable snapshot allocation during shutdown

henningandersen opened the pull request:
During a shutdown, searchable snapshots should not be re-allocated to new nodes,
since this leads to data download. Instead, now wait for the shutdown reallocation
delay before reallocating shards.

Fixes #85052

henningandersen committed:
Delay searchable snapshot allocation during shutdown

During a shutdown, searchable snapshots should not be re-allocated to new nodes,
since this leads to data download. Instead, now wait for the shutdown reallocation
delay before reallocating shards.

Fixes #85052

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

henningandersen committed:
Update docs/changelog/86153.yaml

elasticsearchmachine commented:
Hi @henningandersen, I've created a changelog YAML for you.

henningandersen committed:
Fix changelog yaml

henningandersen committed:
spotless

tlrx commented on x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotAllocator.java:
```diff
@@ -341,11 +342,26 @@ private AllocateUnassignedDecision decideAllocation(RoutingAllocation allocation
         return AllocateUnassignedDecision.NOT_TAKEN;
     }
 
+    private boolean isDelayedDueToShutdown(RoutingAllocation allocation, ShardRouting shardRouting) {
```
```suggestion
    private boolean isDelayedDueToNodeRestart(RoutingAllocation allocation, ShardRouting shardRouting) {
```

henningandersen replied:
Forgot this in prior round, done now in df9115a0833a8344c9a1c38927b0029b10a2d896

tlrx commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotShutdownIntegTests.java:
```diff
@@ -0,0 +1,116 @@
+                }
+            });
+            // leave shutdown in place for some nodes to check that shards get assigned anyway.
+            if (randomBoolean()) {
```
I wonder if this can cause some test failures when a low number of data nodes is picked up? I quickly tested with a 10 shards index and 2 data nodes and got a failure after ~5 runs.

tlrx reviewed (commented):
This looks good but I suspect that the test will cause failures on CI

henningandersen committed:
Make test not fail

Test could fail if a shard received no docs, since then the shard has no size in the
searchable snapshot cache. Now create multiple indices instead of a single index with
multiple shards. Also wait for download activity to cease.

tlrx commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotShutdownIntegTests.java:
```diff
@@ -0,0 +1,134 @@
+
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, minNumDataNodes = 2, numDataNodes = 2)
```
```suggestion
@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, minNumDataNodes = 2)
```

henningandersen replied:
:thu:, leftover from some testing, done in 263b644ec66c3e1fff7031bf255e3c3904c51d52

tlrx reviewed (approved):
LGTM - I left some minor comments.

henningandersen committed:
Rename is delayed method.

henningandersen committed:
Allow more data nodes

henningandersen committed:
Only one repo and better index name

henningandersen committed:
Move number of shards method

henningandersen committed:
Different snapshot names

henningandersen commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotShutdownIntegTests.java:
```diff
@@ -0,0 +1,134 @@
+        int count = between(1, 10);
+        List<String> restoredIndices = new ArrayList<>();
+        for (int i = 0; i < count; ++i) {
+            final String indexName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
```
:+1: , see 9db957e6cc136ebef03d4c5920723ae158bf2851, also now just does one repo.

henningandersen commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotShutdownIntegTests.java:
```diff
@@ -0,0 +1,134 @@
+            final String indexName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
+            createAndPopulateIndex(indexName, Settings.builder());
+
+            final String repositoryName = randomAlphaOfLength(10).toLowerCase(Locale.ROOT);
```
Changed to one repo in 9db957e6cc136ebef03d4c5920723ae158bf2851

henningandersen commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotShutdownIntegTests.java:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    @Override
+    protected int numberOfShards() {
```
:+1: , see 9db957e6cc136ebef03d4c5920723ae158bf2851
